File: sdk/typescript/src/api/browse.ts
import type { HttpTransport } from '../transport.js';
import type { Entry } from '../types.js';
import { isDirectory } from '../types.js';

export class BrowseApi {
  constructor(private readonly http: HttpTransport) {}

  async entries(path?: string): Promise<Entry[]> {
    const data = await this.http.execute<{ treeGetEntries: Entry[] }>(/* GraphQL */ `
      query Browse($path: String) {
        treeGetEntries(path: $path) { name attr timeWrite customaction }
      }
    `, { path });
    return data.treeGetEntries;
  }

  async directories(path?: string): Promise<Entry[]> {
    const all = await this.entries(path);
    return all.filter(isDirectory);
  }

  async files(path?: string): Promise<Entry[]> {
    const all = await this.entries(path);
    return all.filter((e) => !isDirectory(e));
  }
}

File: sdk/python/src/rockbox_sdk/api/browse.py
"""Filesystem browser (also surfaces UPnP entries)."""

from __future__ import annotations

from ..transport import HttpTransport
from ..types import Entry


class BrowseApi:
    def __init__(self, http: HttpTransport) -> None:
        self._http = http

    async def entries(self, path: str | None = None) -> list[Entry]:
        data = await self._http.execute(
            "query Browse($path: String) "
            "{ treeGetEntries(path: $path) "
            "{ name attr timeWrite customaction displayName } }",
            {"path": path},
        )
        return [Entry.model_validate(e) for e in data.get("treeGetEntries", [])]

    async def directories(self, path: str | None = None) -> list[Entry]:
        return [e for e in await self.entries(path) if e.is_directory]

    async def files(self, path: str | None = None) -> list[Entry]:
        return [e for e in await self.entries(path) if not e.is_directory]
